pipelines/google_sheets/data_validator.py

```diff
@@ -0,0 +1,80 @@
+# This is a helper module that contains function which validate data
+
+from typing import Union
+from dateutil import parser
+from datetime import datetime
+
+import pytest
+
+
+# this function takes an url to a Google spreadsheet and computes the spreadsheet id from it. returns an empty string if the url format is not correct
+# spreadsheet url formula: https://docs.google.com/spreadsheets/d/<spreadsheet_id>/edit
+# @input: url- str
+# @output spreadsheet_id: str
+def process_url(url: str) -> str:
+
+    # this string comes before the id
+    url_id_identifier = "d"
+    # split on the '/'
+    parts = url.split("/")
+
+    # loop through parts
+    for i in range(len(parts)):
+        # if we find
+        if parts[i] == url_id_identifier and i+1 < len(parts):
+            if parts[i+1] == "":
+                raise ValueError("Spreadsheet ID is an empty string")
+            return parts[i+1]
+    # if url cannot be found, raise error
+    raise ValueError("Invalid URL. Cannot find spreadsheet ID")
+
+
+# This function receives a serial number which can be an int or float(depending on the serial number) and outputs a datetime object
+# @input: serial_number- int/float. The integer part shows the number of days since December 30th 1899, the decimal part shows the fraction of the day
+# @output: converted_date: datetime object for the same date as the serial number
+def serial_date_to_datetime(serial_number: Union[int, float]) -> datetime:
+
+    # convert starting date to ordinal date: number of days since Jan 1st, 1 AD
+    days_since = datetime(1899, 12, 30).toordinal() + serial_number
```

Comment:
do not create this constant date every time function is called. it is a waste of cpu cycles.

Reply:
btw. in singer code the epoch is different?
```python
excel_epoch = 25569 # 1970-01-01T00:00:00Z, Lotus Notes Serial Number for Epoch Start Date
```

Reply:
According to the google sheets api documentation, the 1899 date is the correct one. This might be a different Lotus Serial Number format they use. Here is the documentation for google sheets' serial numbers https://developers.google.com/sheets/api/reference/rest/v4/DateTimeRenderOption